python/zfp.pyx

```diff
@@ -0,0 +1,210 @@
+        raise TypeError("Input array cannot be None")
+    num_params_set = sum([1 for x in [tolerance, rate, precision] if x > -1])
+    if num_params_set == 0:
+        raise ValueError("Either tolerance, rate, or precision must be set")
+    elif num_params_set > 1:
+        raise ValueError("Only one of tolerance, rate, or precision can be set")
+
+    # Setup zfp structs to begin compression
+    cdef zfp_field* field = _init_field(arr)
+    stream = zfp_stream_open(NULL)
+
+    cdef zfp_type ztype = zfp_type_none;
+    cdef int ndim = arr.ndim;
+    if tolerance > -1:
+        zfp_stream_set_accuracy(stream, tolerance)
+    elif rate > -1:
+        ztype = dtype_to_ztype(arr.dtype)
+        zfp_stream_set_rate(stream, rate, ztype, ndim, 0)
+    elif precision > -1:
+        zfp_stream_set_precision(stream, precision)
+
+    # Allocate space based on the maximum size potentially required by zfp to
+    # store the compressed array
+    cdef bytes compress_str = None
+    cdef size_t maxsize = zfp_stream_maximum_size(stream, field)
+    with Memory(maxsize) as data:
+        bstream = stream_open(data, maxsize)
+        zfp_stream_set_bit_stream(stream, bstream)
+        zfp_stream_rewind(stream)
+        # write the full header so we can reconstruct the numpy array on
+        # decompression
+        if write_header:
+            zfp_write_header(stream, field, HEADER_FULL)
+        compressed_size = zfp_compress(stream, field)
+        # copy the compressed data into a perfectly sized bytes object
+        compress_str = (<char *>data)[:compressed_size]
+
+    zfp_field_free(field)
+    zfp_stream_close(stream)
+    stream_close(bstream)
+
+    return compress_str
+
+cdef np.ndarray _decompress_with_view(zfp_field* field, zfp_stream* stream):
+    cdef zfp_type ztype = field[0]._type
+    dtype = ztype_to_dtype(ztype)
+    format_type = dtype_to_format(dtype)
+
+    shape = (field[0].nx, field[0].ny, field[0].nz, field[0].nw)
+    shape = tuple([x for x in shape if x > 0])
+    num_elements = functools.reduce(operator.mul, shape)
+
+    cdef view.array decomp_arr = view.array(shape, itemsize=sizeof_ztype(ztype),
+                                            format=format_type, mode="fortran",
+                                            allocate_buffer=True)
+    cdef void* pointer = <void *> decomp_arr.data
+    zfp_field_set_pointer(field, pointer)
+    if zfp_decompress(stream, field) == 0:
+        raise RuntimeError("error during zfp decompression")
+
+    return np.asarray(decomp_arr)
+
+cpdef np.ndarray decompress_numpy(bytes compressed_data):
+    if compressed_data is None:
+        raise TypeError
+
+    cdef char* comp_data_pointer = compressed_data
+    cdef bitstream* bstream = stream_open(comp_data_pointer, len(compressed_data))
+    cdef zfp_field* field = zfp_field_alloc()
+    cdef zfp_stream* stream = zfp_stream_open(bstream)
+
+    zfp_stream_rewind(stream)
+    zfp_read_header(stream, field, HEADER_FULL)
```

Comment:
This assumes that compress_numpy wrote a header or that the compressed data came from somewhere else.  At the very least, the return value should be checked.  If no header (or less than a full header) was stored, then this function will fail and return zero.

The documentation needs to make it clear that decompress_numpy requires a header.

Reply:
In the process of supporting @salasoom's new test utils checksum functions, this whole function got overhauled.  The decompression now only requires either the header or the user to provide the metadata (compression parameters, shape, strides, and type).  If both are provided, then they are compared against one another and an exception is raised if they differ (e.g., the header has `float` as the type but the user specified `int32`).